pennylane/transforms/hamiltonian_expand.py

```diff
@@ -172,7 +172,12 @@ def processing_fn(res_groupings):
                     qml.math.dot(r_group, c_group)
                     for c_group, r_group in zip(coeff_groupings, res_groupings)
                 ]
-            return qml.math.sum(qml.math.stack(dot_products), axis=0)
+            summed_dot_products = qml.math.sum(qml.math.stack(dot_products), axis=0)
+            return (
+                qml.math.convert_like(summed_dot_products, res_groupings[0])
+                if len(res_groupings) > 0
```

Comment:
Is there a case where `len(res_groupings) = 0`?

Reply:
Not sure actually, but wanted to make sure we can index `res_groupings` because I think this is the first point in the function where the code could be broken if by any chance an empty sequence was passed as `res_groupings`. Could be wrong and if we want I can investigate this further.

Reply:
If `len(res_groupings) = 0` then `dot_products = []`, and `qml.math.stack(dot_products)` would break.

Reply:
Given that `hamiltonian_expand` only works when returning the expval of a Hamiltonian, I believe we will always have a value returned, and thus `len(res_groupings) > 0`

Reply:
Same concern as @AlbertMitjans . For one, we need to unit test both cases, and therefore understand where the one case would come from.

Reply:
Created https://github.com/PennyLaneAI/pennylane/pull/3411/, this is a responsibility that doesn't belong in this PR.

Overall, if we have Hamiltonian terms, we need to have `res_groupings` because we can assume that `res_groupings` is the output of running the tapes generated by this transform.